### PR TITLE
Search providers

### PR DESCRIPTION
**Pull request by grandpaslab**
Move search to provider architecture.

### CODE REVIEW

**Comment by kemitche** [comment]
Can you rebase this PR appropriately? It appears that some merge commits are dirtying the diff. It's hard to review since there are unrelated changes in there.

**Comment by grandpaslab** [comment]
Sorry about that.  I'm a terrible noob at git. I'll sort through it.

File: r2/r2/lib/providers/search/cloudsearch.py [hunk 1]
```diff
@@ -41,7 +40,9 @@
 from r2.models import (Account, Link, Subreddit, Thing, All, DefaultSR,
                        MultiReddit, DomainSR, Friends, ModContribSR,
                        FakeSubreddit, NotFound)
-
+from r2.lib.providers.search import SearchProvider
+from r2.lib.providers.search.base import safe_get, InvalidQuery, SearchHTTPError, \
```
Comment (kemitche) [inline r2/r2/lib/providers/search/cloudsearch.py]: Small nit: Our preferred format for long import lines is:

```
from a.b.x import (
    alpha,
    beta,
    gamma,
)
``

See also reddit/styleguide

File: r2/r2/lib/providers/search/solrsearch.py [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+from r2.lib.configparse import ConfigValue
+import r2.lib.utils as r2utils
+from r2.models import (Account, Link, Subreddit, All, DefaultSR,
+                       MultiReddit, DomainSR, Friends, ModContribSR,
+                       FakeSubreddit, NotFound)
+from pylons import g, c
+
+from r2.lib.providers.search.base import safe_get, InvalidQuery, SearchHTTPError, \
+        LinkFields, SubredditFields, Results
+from r2.lib.providers.search import SearchProvider
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Please order the imports according to PEP 8 and our styleguide: https://github.com/reddit/styleguide/tree/master/python

Specifically, standard lib imports first, then a blank line, then 3rd party imports (such as pylons), then a blank line, then r2 imports. Each group should be alphabetized.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+
+
+_CHUNK_SIZE = 4000000 # Approx. 4 MB, to stay under the 5MB limit
+_VERSION_OFFSET = 13257906857
+ILLEGAL_XML = re.compile(u'[\x00-\x08\x0b\x0c\x0e-\x1F\uD800-\uDFFF\uFFFE\uFFFF]')
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: ILLEGAL_XML looks unused, if that's correct, please remove it.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+
+
+_SEARCH = "/solr/select?"
+INVALID_QUERY_CODES = ('CS-UnknownFieldInMatchExpression',
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Are these actually used by Solr?

File: r2/r2/lib/providers/search/solrsearch.py [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+                start=0, rank="", return_fields=None, record_stats=False,
+                search_api=None):
+    if search_api is None:
+        search_api = g.solr_search_host
+    if faceting is None:
+        faceting = DEFAULT_FACETS
+    path = _encode_query(query, faceting, size, start, rank, return_fields)
+    timer = None
+    if record_stats:
+        timer = g.stats.get_timer("cloudsearch_timer")
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Should rename the timer here!

File: r2/r2/controllers/front.py [hunk 6]
```diff
@@ -866,7 +850,7 @@ def GET_related(self, num, article, after, reverse, count):
         end = int(time_module.mktime((article._date + rel_range).utctimetuple()))
         nsfw = u"nsfw:0" if not (article.over_18 or article._nsfw.findall(article.title)) else u""
         query = u"(and %s timestamp:%s..%s %s)" % (query, start, end, nsfw)
```
Comment (kemitche) [inline r2/r2/controllers/front.py]: The construction of this "related" query probably needs to be different based on the supported syntax(es) of the provider.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+    sorts_menu_mapping = {}
+    recents = {None: None}
+    known_syntaxes = ("solr", "plain")
+    default_syntax = "solr"
+
+    def __init__(self, query, sr=None, sort=None, syntax=None, raw_sort=None,
+                 faceting=None, recent=None):
+        if syntax is None:
+            syntax = self.default_syntax
+        elif(syntax == 'cloudsearch'):
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: I don't believe that "cloudsearch" syntax is compatible with Solr; unless that's true, this block should be removed.

**Comment by kemitche** [comment]
git is fun, once when you get used to it! If you have any trouble figuring
out the rebase, just ask.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 8]
```diff
@@ -0,0 +1,742 @@
...
+            return None
+        elif isinstance(sr, MultiReddit):
+            for sr_id in sr.sr_ids:
+                bq.append("sr_id:%s" % sr_id)
+        elif isinstance(sr, DomainSR):
+            bq = ["site:'%s'" % sr.domain]
+        elif sr == Friends:
+            if not c.user_is_loggedin or not c.user.friends:
+                return None
+            # The query limit is roughly 8k bytes. Limit to 200 friends to
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Do Solr instances have a size limit for queries? Is it the same as Cloudsearch's (8k)?

File: r2/r2/lib/providers/search/solrsearch.py [hunk 9]
```diff
@@ -0,0 +1,742 @@
...
+class SolrSearchUploader(object):
+    
+    def __init__(self, solr_host=None, solr_port=None, fullnames=None, version_offset=_VERSION_OFFSET):
+        self.solr_host = solr_host or g.solr_doc_host
+        self.solr_port = solr_port or g.solr_port
+        self._version_offset = version_offset
+        self.fullnames = fullnames    
+
+    def add_xml(self, thing):
+        from r2.lib.providers.search.base import safe_xml_str
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: This should be import-able from the top of the file.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+
+        #add = etree.Element("add", id=thing._fullname)
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Please remove unused code.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 11]
```diff
@@ -0,0 +1,742 @@
...
+                response = etree.fromstring(record)
+                status = response.find(STATUS_XPATH).text
+                if status == '0':
+                    # success! 
+                    adds += len(batch.findall('doc'))
+                    deletes += len(batch.findall('delete'))
+                    for w in response.find(WARNING_XPATH) or []:
+                        warnings.append(w.text)
+
+            g.stats.simple_event("cloudsearch.uploads.adds", delta=adds)
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Please name these stats events more appropriately for the provider :wink:

File: r2/r2/lib/providers/search/solrsearch.py [hunk 12]
```diff
@@ -0,0 +1,742 @@
...
+        # etree magic simultaneously removes the elements from one tree
+        # when they are appended to a different tree
+        right_half.append(xml[half:])
+        for chunk in chunk_xml(left_half, depth=depth):
+            yield chunk
+        for chunk in chunk_xml(right_half, depth=depth):
+            yield chunk
+
+
+@g.stats.amqp_processor('cloudsearch_q')
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: stats rename please!

File: r2/r2/lib/providers/search/solrsearch.py [hunk 13]
```diff
@@ -0,0 +1,742 @@
...
+    link_fns = SolrLinkUploader.desired_fullnames(changed)
+    sr_fns = SolrSubredditUploader.desired_fullnames(changed)
+
+    link_uploader = SolrLinkUploader(g.solr_doc_host, fullnames=link_fns)
+    subreddit_uploader = SolrSubredditUploader(g.solr_subreddit_doc_host,
+                                           fullnames=sr_fns)
+
+    link_time = link_uploader.inject()
+    subreddit_time = subreddit_uploader.inject()
+    cloudsearch_time = link_time + subreddit_time
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: minor rename

File: r2/r2/lib/providers/search/solrsearch.py [hunk 14]
```diff
@@ -0,0 +1,742 @@
...
+
+    def should_index(self, thing):
+        return getattr(thing, 'author_id', None) != -1
+
+ 
+def _progress_key(item):
+    return "%s/%s" % (item._id, item._date)
+
+
+_REBUILD_INDEX_CACHE_KEY = "cloudsearch_cursor_%s"
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: another minor rename

File: r2/r2/lib/providers/search/solrsearch.py [hunk 15]
```diff
@@ -0,0 +1,742 @@
...
+    
+    sorts = LinkSearchQuery.sorts_menu_mapping
+
+    def run_changed(drain=False, min_size=int(getattr(g, 'solr_min_batch', 500)), limit=1000, sleep_time=10, 
+            use_safe_get=False, verbose=False):
+        '''Run by `cron` (through `paster run`) on a schedule to send Things to Solr
+        '''
+        if use_safe_get:
+            SolrSearchUploader.use_safe_get = True
+        amqp.handle_items('cloudsearch_changes', _run_changed, min_size=min_size,
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: Just pointing out that this is one that should NOT change (so you did the right thing by leaving it). An unfortunate name choice on my part; perhaps I'll look into renaming it across the board later.

File: upstart/reddit-consumer-search_q.conf [hunk 16]
```diff
@@ -10,5 +10,5 @@ respawn limit 10 5
 nice 10
 script
     . /etc/default/reddit
-    wrap-job paster run --proctitle cloudsearch_q$x $REDDIT_INI $REDDIT_ROOT/r2/lib/cloudsearch.py -c 'run_changed()'
+    wrap-job paster run --proctitle cloudsearch_q$x $REDDIT_INI -c 'from pylons import g; g.search.run_changed()'
```
Comment (kemitche) [inline upstart/reddit-consumer-search_q.conf]: It should be safe to change `--proctitle cloudsearch_q$x` to `--proctitle search_q$x`.

@rram: The `proctitle` is just informational, right?

**Comment by kemitche** [comment]
:nail_care: (I have finished this review pass; please make changes and let me know with the `:haircut:` emoji)

More info: https://github.com/reddit/reddit/wiki/The-Code-Salon

**Comment by grandpaslab** [comment]
:haircut: Cleaned up the issues you found. Let me know what's next!

**Comment by kemitche** [comment]
Still need a rebase - there should be just one commit in the final set, that I'll run some tests on prior to merging & deploying.

**Comment by kemitche** [comment]
Oh, and a :nail_care:. Almost there!

**Comment by grandpaslab** [comment]
:haircut: Finally starting to grok the git, but I'm afraid I totally failed to roll up my commits. Let me know if you need anything changed, or re-branched or whatever.

**Comment by grandpaslab** [comment]
Working on a proper rebase, should be done this morning

**Comment by kemitche** [comment]
Looks good! I'll start running some tests locally once the rebase is ready. Please feel free to reach out if you need help with that!

**Comment by grandpaslab** [comment]
:haircut: Rebased!

**Comment by kemitche** [comment]
## Update on testing from today:

### Bug: Search providers were not registered in `setup.py`.

Please include the following lines:

```
diff --git a/r2/setup.py b/r2/setup.py
index 9ccea33..f0691dc 100644
--- a/r2/setup.py
+++ b/r2/setup.py
@@ -125,5 +125,8 @@ setup(
     [r2.provider.auth]
     cookie = r2.lib.providers.auth.cookie:CookieAuthenticationProvider
     http = r2.lib.providers.auth.http:HttpAuthenticationProvider
+    [r2.provider.search]
+    cloudsearch = r2.lib.providers.search.cloudsearch:CloudSearchProvider
+    solr = r2.lib.providers.search.solrsearch:SolrSearchProvider
     """,
 )
```

After adding these lines, a reddit developer can run `python setup.py develop` to be able to use the search provider system.

### Bug: `SolrSearchQuery.__repr__` references `bq` before assignment.

Please initialize `self.bq = None` in `SolrSearchQuery.__init__`

### Request: Additional set-up information for Solr.

I've never used Solr before, and I'm having some trouble getting the configuration right for proper testing beyond the above. So far, I've done the following in a standard reddit development environment (Ubuntu 12, with an existing reddit setup configured via `install-reddit.sh`):

1. `sudo apt-get install solr-tomcat`
1. `sudo rm /etc/solr/schema.xml; sudo ln -s ~/src/reddit/solr/schema.xml /etc/solr/schema.xml`
1. Edit `/var/lib/tomcat6/conf/server.xml` to replace use of port 8080 (which is already in use by reddit) with port 8123.
1. `sudo service tomcat6 restart` to apply the above changes.

After the above, I can look at `http://reddit.local:8123/solr` and see that it's *trying* to start up, but I end up with the following error:

```
Severe errors in solr configuration. Check your log files for more detailed information on what may be wrong. If you want solr to continue after configuration errors, change: <abortOnConfigurationError>false</abortOnConfigurationError> in null ------------------------------------------------------------- org.apache.solr.common.SolrException: Schema Parsing Failed at org.apache.solr.schema.IndexSchema.readSchema(IndexSchema.java:656) at org.apache.solr.schema.IndexSchema.<init>(IndexSchema.java:95) at org.apache.solr.core.SolrCore.<init>(SolrCore.java:521) at org.apache.solr.core.CoreContainer$Initializer.initialize(CoreContainer.java:137) at org.apache.solr.servlet.SolrDispatchFilter.init(SolrDispatchFilter.java:83) at org.apache.catalina.core.ApplicationFilterConfig.getFilter(ApplicationFilterConfig.java:295) at org.apache.catalina.core.ApplicationFilterConfig.setFilterDef(ApplicationFilterConfig.java:422) at org.apache.catalina.core.ApplicationFilterConfig.<init>(ApplicationFilterConfig.java:115) at org.apache.catalina.core.StandardContext.filterStart(StandardContext.java:4072) at org.apache.catalina.core.StandardContext.start(StandardContext.java:4726) at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:799) at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:779) at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:601) at org.apache.catalina.startup.HostConfig.deployDescriptor(HostConfig.java:675) at org.apache.catalina.startup.HostConfig.deployDescriptors(HostConfig.java:601) at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:502) at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1317) at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:324) at org.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:142) at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1065) at org.apache.catalina.core.StandardHost.start(StandardHost.java:840) at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1057) at org.apache.catalina.core.StandardEngine.start(StandardEngine.java:463) at org.apache.catalina.core.StandardService.start(StandardService.java:525) at org.apache.catalina.core.StandardServer.start(StandardServer.java:754) at org.apache.catalina.startup.Catalina.start(Catalina.java:595) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:606) at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:289) at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:414) Caused by: java.lang.RuntimeException: unknown field 'id' at org.apache.solr.schema.IndexSchema.getIndexedField(IndexSchema.java:314) at org.apache.solr.schema.IndexSchema.readSchema(IndexSchema.java:602) ... 31 more ------------------------------------------------------------- java.lang.RuntimeException: unknown field 'id' at org.apache.solr.schema.IndexSchema.getIndexedField(IndexSchema.java:314) at org.apache.solr.schema.IndexSchema.readSchema(IndexSchema.java:602) at org.apache.solr.schema.IndexSchema.<init>(IndexSchema.java:95) at org.apache.solr.core.SolrCore.<init>(SolrCore.java:521) at org.apache.solr.core.CoreContainer$Initializer.initialize(CoreContainer.java:137) at org.apache.solr.servlet.SolrDispatchFilter.init(SolrDispatchFilter.java:83) at org.apache.catalina.core.ApplicationFilterConfig.getFilter(ApplicationFilterConfig.java:295) at org.apache.catalina.core.ApplicationFilterConfig.setFilterDef(ApplicationFilterConfig.java:422) at org.apache.catalina.core.ApplicationFilterConfig.<init>(ApplicationFilterConfig.java:115) at org.apache.catalina.core.StandardContext.filterStart(StandardContext.java:4072) at org.apache.catalina.core.StandardContext.start(StandardContext.java:4726) at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:799) at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:779) at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:601) at org.apache.catalina.startup.HostConfig.deployDescriptor(HostConfig.java:675) at org.apache.catalina.startup.HostConfig.deployDescriptors(HostConfig.java:601) at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:502) at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1317) at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:324) at org.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:142) at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1065) at org.apache.catalina.core.StandardHost.start(StandardHost.java:840) at org.apache.catalina.core.ContainerBase.start(ContainerBase.java:1057) at org.apache.catalina.core.StandardEngine.start(StandardEngine.java:463) at org.apache.catalina.core.StandardService.start(StandardService.java:525) at org.apache.catalina.core.StandardServer.start(StandardServer.java:754) at org.apache.catalina.startup.Catalina.start(Catalina.java:595) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:606) at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:289) at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:414)
```

**Comment by kemitche** [comment]
For the record - having step-by-step set-up instructions for configuring a Solr instance provider on a "default" reddit install would be a great thing to drop into the `{reddit}/solr` directory!

**Comment by kemitche** [comment]
@grandpaslab I've narrowed the issue down to some problems with your `schema.xml`. I'm know basically nothing about solr schemas, but from what I can tell, at a minimum, the schema is missing `<fields> ... </fields>` around the field declarations.

If you could review your schema and fix it up such that a Solr instance is able to consume it without errors, I can then continue testing. Please also let me know which version of Solr you're using so I can ensure I'm matching your configuration. Super-extra-bonus points if the schema is consumable by Solr 1.4 as well as whatever version of Solr you're targeting, as Solr 1.4 is what is packaged in the `solr-common` apt package on Ubuntu 12 (reddit's current "default" environment).

**Comment by kemitche** [comment]
:nail_care: (though I'm going to keep poking at the schema myself to see if I can identify further issues)

**Comment by grandpaslab** [comment]
I've just pushed the current schema. FYI I'm running Solr 4.10.3

**Comment by xiongchiamiov** [comment]
We [mostly support Ubuntu 12.04](https://github.com/reddit/reddit/blob/master/install-reddit.sh#L81), which [appears to use](http://packages.ubuntu.com/search?keywords=solr&searchon=names&suite=precise&section=all) solr 1.4.1.  It's not surprising, then, that there are config issues. :)

**Comment by grandpaslab** [comment]
Yeah, I'm on a corporate intranet, so I takes what they gives me, which is CentOS. I'll see about cooking up a config for solr 1.4.1

**Comment by kemitche** [comment]
I have a schema based on munging your schema with the "sample" schema installed on ubuntu 12. I've gotten the solr server to start up with it; I'll need to test more to see if I can use it to upload/search, but progress is being made.

I think there's a strong chance of having a backwards compatible schema that we can use out of the box on ubuntu 12!

**Comment by kemitche** [comment]
```
In [3]: query = g.search.SearchQuery("potato"); query.run()
Out[3]: Results([], 0, {})
```

Got a query off :+1:. I did have to disable the line about "edismax", and I don't have any data in the index. But in short: testing continues.

File: r2/r2/controllers/front.py [hunk 17]
```diff
@@ -853,7 +851,7 @@ def GET_related(self, num, article, after, reverse, count):
         end = int(time_module.mktime((article._date + rel_range).utctimetuple()))
         nsfw = u"nsfw:0" if not article.is_nsfw else u""
         query = u"(and %s timestamp:%s..%s %s)" % (query, start, end, nsfw)
-        q = SearchQuery(query, raw_sort="-text_relevance", faceting={},
+        q = g.search.SearchQuery(query, raw_sort="-text_relevance", faceting={},
```
Comment (kemitche) [inline r2/r2/controllers/front.py]: This still needs to be updated to use `g.search.get_related_query`

File: r2/r2/lib/providers/search/solrsearch.py [hunk 18]
```diff
@@ -0,0 +1,812 @@
...
+    '''Consume the cloudsearch_changes queue, and print reporting information
+    on how long it took and how many remain
+    
+    '''
+    start = datetime.now(g.tz)
+
+    changed = [pickle.loads(msg.body) for msg in msgs]
+
+    link_fns = SolrLinkUploader.desired_fullnames(changed)
+    sr_fns = SolrSubredditUploader.desired_fullnames(changed)
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: `desired_fullnames(...)` isn't defined on these classes.

File: r2/r2/lib/providers/search/cloudsearch.py [hunk 19]
```diff
@@ -1091,3 +797,38 @@ def customize_query(self, bq=u''):
...
+    SearchException = (SearchHTTPError,)
+
+    SearchQuery = LinkSearchQuery
+
+    SubredditSearchQuery = CloudSearchSubredditSearchQuery
+
+    sorts = LinkSearchQuery.sorts_menu_mapping
+
+    def run_changed(drain=False, min_size=int(getattr(g, 'SOLR_MIN_BATCH', 500)), limit=1000, sleep_time=10, 
+            use_safe_get=False, verbose=False):
```
Comment (kemitche) [inline r2/r2/lib/providers/search/cloudsearch.py]: this method (and its partner in solrsearch) also needs to take `self` as its first argument.

File: r2/r2/lib/providers/search/solrsearch.py [hunk 20]
```diff
@@ -0,0 +1,812 @@
...
+        self.srs = {}
+
+    def fields(self, thing):
+        '''Return fields relevant to a Link search index'''
+        account = self.accounts[thing.author_id]
+        sr = self.srs[thing.sr_id]
+        return LinkFields(thing, account, sr).fields()
+
+    def batch_lookups(self):
+        super(SolrLinkUploader, self).batch_lookups()
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solrsearch.py]: There's no `batch_lookups()` function on the super class, so this line is failing.

File: r2/r2/controllers/front.py (hunk 17 above)
Comment (grandpaslab) [inline r2/r2/controllers/front.py]: Sorry, lost those lines as a result of yesterday's hard reset. Should be good now.

**Comment by grandpaslab** [comment]
:haircut: Fixes are in for these. I'm at home today, not able to fully test due to connectivity issues through our firewall/proxy, but the changes are all pretty straightforward.

**Comment by xiongchiamiov** [comment]
:runner:

**Comment by grandpaslab** [comment]
I've got Solr 1.4.1 running, but it looks like some code changes are needed. Stay tuned.

**Comment by grandpaslab** [comment]
:haircut: Support for Solr 1.x and 4.x is in. Only URLs are different, so change was minor, with a new 'solr_version' config option to choose URL structure. Default is 1. I've also started a README. Since I don't have Ubuntu handy I'm hoping you can confirm it's correct / suggest changes.

**Comment by kemitche** [comment]
Just FYI: I'm a little swamped at the moment, but should be able to
continue testing & review on this early next week!

**Comment by grandpaslab** [comment]
:haircut: ping

**Comment by kemitche** [comment]
I'm looking at this today! Some quick notes: I recently refactored r2.lib.amqp a bit so that it could be imported prior to pylons.g being set-up. That means, I think, that you can set up `self.search_provider` with the rest of the providers in app_globals.py (though feel free to leave `g.search` as a property that access `self.search_provider`. You may still see an import loop cycle due to the `epoch_seconds` import from `r2.lib.db.sorts`; you can get around that by using `r2.lib.utils.epoch_timestamp`.

**Comment by kemitche** [comment]
Note: I'm still testing though, so don't run off and make those changes. The models code also does early imports of `g` which means that search providers may still need that delayed load as you initially implemented it.

File: r2/r2/lib/menus.py [hunk 21]
```diff
@@ -609,7 +605,7 @@ def make_title(self, attr):
 class SearchSortMenu(SortMenu):
     """Sort menu for search pages."""
     _default = 'relevance'
-    mapping = search_sorts
+    mapping = g.search.sorts
```
Comment (kemitche) [inline r2/r2/lib/menus.py]: This line is causing problems; since class variables like this get created at module import time, it's trying to read from `g` at import time, which can be before pylons has configured `g`. Even if it works now or in your environment, all it takes is a change in import ordering, or a plugin importing things differently, to trigger unresolvable `AttributeError`s on app start up.

Looking at the other code: Nothing appears to rely on `mapping` or `_options` being available at the class level, so these should be defined as `property`s instead, so that `g.search.sorts` is read at run-time rather than at import time.

File: r2/r2/lib/menus.py [hunk 22]
```diff
@@ -620,7 +616,7 @@ def operator(cls, sort):
 class SubredditSearchSortMenu(SortMenu):
     """Sort menu for subreddit search pages."""
     _default = 'relevance'
-    mapping = sr_search_sorts
+    mapping = g.search.sorts
     _options = mapping.keys()
```
Comment (kemitche) [inline r2/r2/lib/menus.py]: Same note here about `mapping` and `_options`.

File: r2/r2/lib/menus.py (hunk 21 above)
Comment (kemitche) [inline r2/r2/lib/menus.py]: Note: I realize that it *appears* that `cls.mapping` is used below; but from what I can tell, that is dead code; there's no `cls.default`\* defined so that code would crash.

\* `cls._default`, with an underscore *is* defined and used elsewhere.

File: r2/r2/controllers/front.py [hunk 23]
```diff
@@ -955,7 +944,7 @@ def GET_search_reddits(self, query, reverse, after, count, num):
               recent=VMenu('t', TimeMenu, remember=False),
               restrict_sr=VBoolean('restrict_sr', default=False),
               include_facets=VBoolean('include_facets', default=False),
-              syntax=VOneOf('syntax', options=SearchQuery.known_syntaxes))
+              syntax=VOneOf('syntax', options=g.search.SearchQuery.known_syntaxes))
```
Comment (kemitche) [inline r2/r2/controllers/front.py]: Here's another tricky case of reading something of `g` at import time.

I'm testing locally a separate validator for this that defers reading the choices off of `g` until runtime:

```python
class VSearchSyntax(VOneOf):
    def run(self, val):
        # Deferred setting of self.options to avoid reading from pylons.g
        # at import time (which is when Validators are instantiated)
        self.options = g.search.SearchQuery.known_syntaxes
        return VOneOf.run(self, val)
```

File: r2/r2/lib/menus.py (hunk 21 above)
Comment (kemitche) [inline r2/r2/lib/menus.py]: I've found that `VMenu.param_docs` relies on looking up `_options` from the class, rather than an instance of the class (which more or less makes sense, since `param_docs` runs at load time.

This of course complicates the situation and means I don't have a good solution to this tangled web of premature access to `pylons.g` just yet.

File: r2/r2/lib/providers/search/__init__.py [hunk 24]
```diff
@@ -20,14 +20,28 @@
...
+        return NotImplementedError
+
+    def sorts(self):
+        return NotImplementedError
+
+    def run_changed(self):
+        return NotImplementedError
+
+    def get_related_query(self):
+        return NotImplementedError
```
Comment (kemitche) [inline r2/r2/lib/providers/search/__init__.py]: All these functions should `raise NotImplementedError` rather than returning it.

(`return NotImplemented` is a pattern more used for implementing comparators, I believe)

**Comment by kemitche** [comment]
End of business day update:

- [x] Able to set-up Solr with instructions given in README
- [ ] Can run reddit code (non-search features) with providerized search (Solr and/or Cloudsearch)
- [ ] Can upload docs to Solr
- [ ] Can search the Solr index
- [ ] Verified no regressions to Cloudsearch functionality used by reddit.com

The second bullet point is causing trouble; I'm hitting cyclical issues with reading `g.search` and attributes off of the provider in various contexts. It's a combination of import-time execution of validators, module level globals reading off of `g`, and similar problems.

Other providers don't hit these issues as they don't import any code that interacts with `g`. It seems like in order to make this providerization work, there's going to need to either be (1) changes to the providers to de-couple those files from most `r2` imports, or (2) changes to things like `VMenu` and such to not read from `g.search` at import time of controllers, etc.

I lean towards option 2 at this point; I've experimented a bit with the path that needs taking, and the big problem is the `VMenu` validator's use of `SearchSortMenu` and `SubredditSearchSortMenu`, which each try to pull information from `g.search` with this pull request. The straightest path forward may be to provide a static set of sorts (`_options`) for the initial implementation. It's not ideal or elegant, but my goal is to get this merged sooner rather than let the PR live too long and cause merge issues.

So, to summarize: If you have time, please *remove the references to `g.search` from `SearchSortMenu` and `SubredditSearchSortMenu`*, and define those mappings/_options globally. We can refactor later to find a way to let each provider define its sort options & mappings.

File: r2/r2/lib/menus.py (hunk 21 above)
Comment (grandpaslab) [inline r2/r2/lib/menus.py]: Trying to get to this today, got some day-job stuff to take care of first.

**Comment by grandpaslab** [comment]
:haircut: I moved the definition of menus to app_globals. Search syntaxes too. Rebased again.

File: r2/r2/lib/providers/search/solr.py [hunk 25]
```diff
@@ -0,0 +1,847 @@
...
+        
+        '''
+
+        xml_things = self.xml_from_things()
+
+        cs_time = 0
+        for batch in xml_things:
+
+            if not len(batch):
+                return 0
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solr.py]: This `return` statement breaks out of the `for` loop, which means if the set of things to `add` is empty, the `delete`s won't get uploaded. It should be a `continue` statement instead.

File: r2/r2/lib/providers/search/solr.py [hunk 26]
```diff
@@ -0,0 +1,847 @@
...
+            else:                                                                                                       
+                raise 
+
+    def xml_from_things(self):
+        '''Generate a <batch> XML tree to send to solr for
+        adding/updating/deleting the given things
+        
+        '''
+        add = etree.Element("add")
+        delete = etree.Element("delete")
```
Comment (kemitche) [inline r2/r2/lib/providers/search/solr.py]: In my testing, I found that the index wasn't ever getting updated with the uploaded results. I determined that the `<commit>` directive isn't ever sent. My recommendation would be to update this function to:

* Set `commit = etree.Element("commit"); commit.attribs["waitSearcher"] = "false"`
* `return (add, delete, commit)` from this function
* Update the loop in `inject()` to not skip an empty `commit` directive (since it will always be a child-less `etree.Element`)

https://cwiki.apache.org/confluence/display/solr/Uploading+Data+with+Index+Handlers

With the above changes, I was able to upload and search locally!

**Comment by kemitche** [comment]
So far so good; other than the minor changes I mentioned a moment ago, we're getting really close. I just need to run some testing of this code when using CloudSearch as the provider to ensure nothing regresses.

**Comment by grandpaslab** [comment]
Nice!  Keep me posted, I shouldn't be too busy for the next couple days.

**Comment by kemitche** [comment]
Heads up for what I'm planning for today: I'm going to rebase your changes and put them on a branch on my fork (https://github.com/kemitche/reddit). That'll give both you and I a new baseline to work off of; I need that so I can perform some testing against cloudsearch scenarios this afternoon.

The branch will also include a commit with the tweaks I mentioned previously that were necessary to reflect changes in the search index.

**Comment by grandpaslab** [comment]
Nice.

**Comment by kemitche** [comment]
Ok, here's my cleaned up branch: https://github.com/kemitche/reddit/tree/search-providers-rebase
(diff, for reference: https://github.com/reddit/reddit/compare/master...kemitche:search-providers-rebase)

You should do the following:

1\. Save your existing branch in an easy to get back to way:

```
$ git checkout search-providers
$ git checkout -b search-providers-backup
```

2\. Pull in my branch and force push it up to github, so that this PR reflects the compacted, cleaned up view:

```
$ git remote add kemitche git@github.com:kemitche/reddit.git
$ git fetch kemitche
$ git push -f origin kemitche/search-providers-rebase:search-providers
```

**Important** to note that the last line is a force push; be careful not to clobber things you don't mean to. In particular, I'm assuming that your "origin" remote is github, but you should modify the command as needed. (Let me know if you need help - and apologies if any of this comes across as condescending, I just recall you mentioning earlier that you're new-ish to git)

It took me a good chunk of my "free" time today to get the rebase done (since there were a lot of merges in your branch that reflected unrelated changes that needed to be cleaned up), so please take advantage of it! I'll try and get some actual testing done tomorrow now that I have a branch with a smaller number of targeted, clean commits.

**Comment by grandpaslab** [comment]
Done. Thanks very much for the cleanup. Extra merges were totally the
result of my git-noobishness.

**Comment by kemitche** [comment]
Merged in as 990a583..00fbdcb!

Thank you for the awesome work. Can you confirm your reddit username for me, so I can grant you an open sorcerer trophy?

**Comment by grandpaslab** [comment]
Thanks for your patience, and thanks for the gold!
